File: platform/Flutter/sweeteditor/lib/keymap/editor_keymap.dart
import '../core/editor_core.dart' as core;

export '../core/editor_core.dart'
    show EditorBuiltinCommand, KeyBinding, KeyChord, KeyModifier, KeyCode;

typedef EditorCommandHandler = bool Function();

class EditorKeyMap {
  EditorKeyMap({Iterable<core.KeyBinding>? bindings}) {
    if (bindings != null) {
      for (final binding in bindings) {
        addBinding(binding);
      }
    }
  }

  final List<core.KeyBinding> _bindings = <core.KeyBinding>[];
  final Map<int, EditorCommandHandler> _handlers =
      <int, EditorCommandHandler>{};
  int _nextCustomCommandId =
      core.EditorBuiltinCommand.triggerCompletion.value + 1;

  List<core.KeyBinding> get bindings =>
      List<core.KeyBinding>.unmodifiable(_bindings);

  void addBinding(core.KeyBinding binding) {
    if (binding.first.keyCode == core.KeyCode.none) return;
    _bindings.removeWhere(
      (existing) =>
          _sameChord(existing.first, binding.first) &&
          _sameChord(existing.second, binding.second),
    );
    _bindings.add(binding);
  }

  void removeBinding(core.KeyBinding binding) {
    _bindings.removeWhere(
      (existing) =>
          _sameChord(existing.first, binding.first) &&
          _sameChord(existing.second, binding.second),
    );
  }

  static bool _sameChord(core.KeyChord a, core.KeyChord b) {
    return a.modifiers == b.modifiers && a.keyCode == b.keyCode;
  }

  factory EditorKeyMap.defaultKeyMap() => EditorKeyMap.vscode();

  static void _bind(
    EditorKeyMap keyMap,
    int modifiers,
    int keyCode,
    int command,
  ) {
    keyMap.addBinding(
      core.KeyBinding(
        first: core.KeyChord(modifiers: modifiers, keyCode: keyCode),
        command: command,
      ),
    );
  }

  static void _addCommonBindings(EditorKeyMap keyMap) {
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.left,
      core.EditorBuiltinCommand.cursorLeft.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.right,
      core.EditorBuiltinCommand.cursorRight.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.up,
      core.EditorBuiltinCommand.cursorUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.down,
      core.EditorBuiltinCommand.cursorDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.home,
      core.EditorBuiltinCommand.cursorLineStart.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.end,
      core.EditorBuiltinCommand.cursorLineEnd.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.pageUp,
      core.EditorBuiltinCommand.cursorPageUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.pageDown,
      core.EditorBuiltinCommand.cursorPageDown.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.left,
      core.EditorBuiltinCommand.selectLeft.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.right,
      core.EditorBuiltinCommand.selectRight.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.up,
      core.EditorBuiltinCommand.selectUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.down,
      core.EditorBuiltinCommand.selectDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.home,
      core.EditorBuiltinCommand.selectLineStart.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.end,
      core.EditorBuiltinCommand.selectLineEnd.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.pageUp,
      core.EditorBuiltinCommand.selectPageUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.pageDown,
      core.EditorBuiltinCommand.selectPageDown.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.backspace,
      core.EditorBuiltinCommand.backspace.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.deleteKey,
      core.EditorBuiltinCommand.deleteForward.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.tab,
      core.EditorBuiltinCommand.insertTab.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.none,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertNewline.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.a,
      core.EditorBuiltinCommand.selectAll.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.a,
      core.EditorBuiltinCommand.selectAll.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.z,
      core.EditorBuiltinCommand.undo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.z,
      core.EditorBuiltinCommand.undo.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.c,
      core.EditorBuiltinCommand.copy.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.c,
      core.EditorBuiltinCommand.copy.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.v,
      core.EditorBuiltinCommand.paste.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.v,
      core.EditorBuiltinCommand.paste.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.x,
      core.EditorBuiltinCommand.cut.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.x,
      core.EditorBuiltinCommand.cut.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.space,
      core.EditorBuiltinCommand.triggerCompletion.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.space,
      core.EditorBuiltinCommand.triggerCompletion.value,
    );
  }

  factory EditorKeyMap.vscode() {
    final keyMap = EditorKeyMap();
    _addCommonBindings(keyMap);

    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.z,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.z,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.y,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.y,
      core.EditorBuiltinCommand.redo.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineBelow.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineBelow.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineAbove.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineAbove.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.alt,
      core.KeyCode.up,
      core.EditorBuiltinCommand.moveLineUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.alt,
      core.KeyCode.down,
      core.EditorBuiltinCommand.moveLineDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.alt | core.KeyModifier.shift,
      core.KeyCode.up,
      core.EditorBuiltinCommand.copyLineUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.alt | core.KeyModifier.shift,
      core.KeyCode.down,
      core.EditorBuiltinCommand.copyLineDown.value,
    );

    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.k,
      core.EditorBuiltinCommand.deleteLine.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.k,
      core.EditorBuiltinCommand.deleteLine.value,
    );

    return keyMap;
  }

  factory EditorKeyMap.jetbrains() {
    final keyMap = EditorKeyMap();
    _addCommonBindings(keyMap);

    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.z,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.z,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.shift,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineBelow.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.alt,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineAbove.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.alt,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineAbove.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.alt | core.KeyModifier.shift,
      core.KeyCode.up,
      core.EditorBuiltinCommand.moveLineUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.alt | core.KeyModifier.shift,
      core.KeyCode.down,
      core.EditorBuiltinCommand.moveLineDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.d,
      core.EditorBuiltinCommand.copyLineDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.d,
      core.EditorBuiltinCommand.copyLineDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.y,
      core.EditorBuiltinCommand.deleteLine.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.y,
      core.EditorBuiltinCommand.deleteLine.value,
    );

    return keyMap;
  }

  factory EditorKeyMap.sublime() {
    final keyMap = EditorKeyMap();
    _addCommonBindings(keyMap);

    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.y,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.z,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.z,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.y,
      core.EditorBuiltinCommand.redo.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineBelow.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineBelow.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineAbove.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.enter,
      core.EditorBuiltinCommand.insertLineAbove.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.up,
      core.EditorBuiltinCommand.moveLineUp.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.down,
      core.EditorBuiltinCommand.moveLineDown.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.ctrl | core.KeyModifier.shift,
      core.KeyCode.k,
      core.EditorBuiltinCommand.deleteLine.value,
    );
    _bind(
      keyMap,
      core.KeyModifier.meta | core.KeyModifier.shift,
      core.KeyCode.k,
      core.EditorBuiltinCommand.deleteLine.value,
    );

    return keyMap;
  }

  int registerCommand(core.KeyBinding binding, EditorCommandHandler handler) {
    var resolvedBinding = binding;
    var command = binding.command;
    if (command == core.EditorBuiltinCommand.none.value) {
      command = _nextCustomCommandId++;
      resolvedBinding = core.KeyBinding(
        first: binding.first,
        second: binding.second,
        command: command,
      );
    } else if (command >= _nextCustomCommandId) {
      _nextCustomCommandId = command + 1;
    }

    addBinding(resolvedBinding);
    _handlers[command] = handler;
    return command;
  }

  bool invokeHandler(int command) => _handlers[command]?.call() ?? false;
}

File: platform/Android/sweeteditor/src/main/java/com/qiplat/sweeteditor/EditorKeyMap.java
package com.qiplat.sweeteditor;

import android.util.SparseArray;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.qiplat.sweeteditor.core.keymap.EditorBuiltinCommand;
import com.qiplat.sweeteditor.core.keymap.KeyBinding;
import com.qiplat.sweeteditor.core.keymap.KeyCode;
import com.qiplat.sweeteditor.core.keymap.KeyModifier;

import java.util.ArrayList;
import java.util.List;

/**
 * Widget-layer keyboard shortcut map that additionally holds command handlers.
 */
public class EditorKeyMap {

    @FunctionalInterface
    public interface ShortcutHandler {
        void onShortcut(@NonNull KeyBinding binding, @NonNull SweetEditor editor);
    }

    private final ArrayList<KeyBinding> mBindings = new ArrayList<>();
    private final SparseArray<ShortcutHandler> mCommands = new SparseArray<>();
    private int mNextCustomId = EditorBuiltinCommand.TRIGGER_COMPLETION.value + 1;

    public void addBinding(@NonNull KeyBinding binding) {
        mBindings.remove(binding);
        mBindings.add(binding);
    }

    public void removeBinding(@NonNull KeyBinding binding) {
        mBindings.remove(binding);
    }

    @NonNull
    public List<KeyBinding> getBindings() {
        return mBindings;
    }

    /**
     * Register a command handler for the given binding.
     * <p>
     * If {@code binding.command == EditorBuiltinCommand.NONE.value}, a custom command id is auto-assigned.
     *
     * @return the resolved command id
     */
    public int registerCommand(@NonNull KeyBinding binding,
                               @NonNull ShortcutHandler handler) {
        int commandId = binding.command;
        KeyBinding resolvedBinding = binding;
        if (commandId == EditorBuiltinCommand.NONE.value) {
            commandId = mNextCustomId++;
            resolvedBinding = new KeyBinding(binding.first, binding.second, commandId);
        } else if (commandId >= mNextCustomId) {
            mNextCustomId = commandId + 1;
        }
        mCommands.put(commandId, handler);
        addBinding(resolvedBinding);
        return commandId;
    }

    @Nullable
    public ShortcutHandler getCommand(int commandId) {
        return mCommands.get(commandId);
    }

    private static void bind(EditorKeyMap km, int modifiers, int keyCode, int command) {
        km.addBinding(new KeyBinding(modifiers, keyCode, command));
    }

    private static void addCommonBindings(EditorKeyMap km) {
        bind(km, KeyModifier.NONE, KeyCode.LEFT,      EditorBuiltinCommand.CURSOR_LEFT.value);
        bind(km, KeyModifier.NONE, KeyCode.RIGHT,     EditorBuiltinCommand.CURSOR_RIGHT.value);
        bind(km, KeyModifier.NONE, KeyCode.UP,        EditorBuiltinCommand.CURSOR_UP.value);
        bind(km, KeyModifier.NONE, KeyCode.DOWN,      EditorBuiltinCommand.CURSOR_DOWN.value);
        bind(km, KeyModifier.NONE, KeyCode.HOME,      EditorBuiltinCommand.CURSOR_LINE_START.value);
        bind(km, KeyModifier.NONE, KeyCode.END,       EditorBuiltinCommand.CURSOR_LINE_END.value);
        bind(km, KeyModifier.NONE, KeyCode.PAGE_UP,   EditorBuiltinCommand.CURSOR_PAGE_UP.value);
        bind(km, KeyModifier.NONE, KeyCode.PAGE_DOWN, EditorBuiltinCommand.CURSOR_PAGE_DOWN.value);

        bind(km, KeyModifier.SHIFT, KeyCode.LEFT,      EditorBuiltinCommand.SELECT_LEFT.value);
        bind(km, KeyModifier.SHIFT, KeyCode.RIGHT,     EditorBuiltinCommand.SELECT_RIGHT.value);
        bind(km, KeyModifier.SHIFT, KeyCode.UP,        EditorBuiltinCommand.SELECT_UP.value);
        bind(km, KeyModifier.SHIFT, KeyCode.DOWN,      EditorBuiltinCommand.SELECT_DOWN.value);
        bind(km, KeyModifier.SHIFT, KeyCode.HOME,      EditorBuiltinCommand.SELECT_LINE_START.value);
        bind(km, KeyModifier.SHIFT, KeyCode.END,       EditorBuiltinCommand.SELECT_LINE_END.value);
        bind(km, KeyModifier.SHIFT, KeyCode.PAGE_UP,   EditorBuiltinCommand.SELECT_PAGE_UP.value);
        bind(km, KeyModifier.SHIFT, KeyCode.PAGE_DOWN, EditorBuiltinCommand.SELECT_PAGE_DOWN.value);

        bind(km, KeyModifier.NONE, KeyCode.BACKSPACE,  EditorBuiltinCommand.BACKSPACE.value);
        bind(km, KeyModifier.NONE, KeyCode.DELETE_KEY,  EditorBuiltinCommand.DELETE_FORWARD.value);
        bind(km, KeyModifier.NONE, KeyCode.TAB,        EditorBuiltinCommand.INSERT_TAB.value);
        bind(km, KeyModifier.NONE, KeyCode.ENTER,      EditorBuiltinCommand.INSERT_NEWLINE.value);

        bind(km, KeyModifier.CTRL, KeyCode.A, EditorBuiltinCommand.SELECT_ALL.value);
        bind(km, KeyModifier.META, KeyCode.A, EditorBuiltinCommand.SELECT_ALL.value);

        bind(km, KeyModifier.CTRL, KeyCode.Z, EditorBuiltinCommand.UNDO.value);
        bind(km, KeyModifier.META, KeyCode.Z, EditorBuiltinCommand.UNDO.value);

        km.registerCommand(
                new KeyBinding(KeyModifier.CTRL, KeyCode.C, EditorBuiltinCommand.COPY.value),
                (binding, editor) -> editor.copyToClipboard());
        km.registerCommand(
                new KeyBinding(KeyModifier.META, KeyCode.C, EditorBuiltinCommand.COPY.value),
                (binding, editor) -> editor.copyToClipboard());
        km.registerCommand(
                new KeyBinding(KeyModifier.CTRL, KeyCode.V, EditorBuiltinCommand.PASTE.value),
                (binding, editor) -> editor.pasteFromClipboard());
        km.registerCommand(
                new KeyBinding(KeyModifier.META, KeyCode.V, EditorBuiltinCommand.PASTE.value),
                (binding, editor) -> editor.pasteFromClipboard());
        km.registerCommand(
                new KeyBinding(KeyModifier.CTRL, KeyCode.X, EditorBuiltinCommand.CUT.value),
                (binding, editor) -> editor.cutToClipboard());
        km.registerCommand(
                new KeyBinding(KeyModifier.META, KeyCode.X, EditorBuiltinCommand.CUT.value),
                (binding, editor) -> editor.cutToClipboard());

        km.registerCommand(
                new KeyBinding(KeyModifier.CTRL, KeyCode.SPACE, EditorBuiltinCommand.TRIGGER_COMPLETION.value),
                (binding, editor) -> editor.triggerCompletion());
        km.registerCommand(
                new KeyBinding(KeyModifier.META, KeyCode.SPACE, EditorBuiltinCommand.TRIGGER_COMPLETION.value),
                (binding, editor) -> editor.triggerCompletion());
    }

    /**
     * Create the default key map (VS Code style).
     */
    public static EditorKeyMap defaultKeyMap() {
        return vscode();
    }

    /**
     * VS Code key bindings.
     */
    public static EditorKeyMap vscode() {
        EditorKeyMap km = new EditorKeyMap();
        addCommonBindings(km);

        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.Z, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.Z, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.CTRL, KeyCode.Y, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.META, KeyCode.Y, EditorBuiltinCommand.REDO.value);

        bind(km, KeyModifier.CTRL, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_BELOW.value);
        bind(km, KeyModifier.META, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_BELOW.value);
        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_ABOVE.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_ABOVE.value);

        bind(km, KeyModifier.ALT, KeyCode.UP,   EditorBuiltinCommand.MOVE_LINE_UP.value);
        bind(km, KeyModifier.ALT, KeyCode.DOWN, EditorBuiltinCommand.MOVE_LINE_DOWN.value);
        bind(km, KeyModifier.ALT | KeyModifier.SHIFT, KeyCode.UP,   EditorBuiltinCommand.COPY_LINE_UP.value);
        bind(km, KeyModifier.ALT | KeyModifier.SHIFT, KeyCode.DOWN, EditorBuiltinCommand.COPY_LINE_DOWN.value);

        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.K, EditorBuiltinCommand.DELETE_LINE.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.K, EditorBuiltinCommand.DELETE_LINE.value);

        return km;
    }

    /**
     * JetBrains (IntelliJ IDEA) key bindings.
     */
    public static EditorKeyMap jetbrains() {
        EditorKeyMap km = new EditorKeyMap();
        addCommonBindings(km);

        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.Z, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.Z, EditorBuiltinCommand.REDO.value);

        bind(km, KeyModifier.CTRL, KeyCode.Y, EditorBuiltinCommand.DELETE_LINE.value);
        bind(km, KeyModifier.META, KeyCode.Y, EditorBuiltinCommand.DELETE_LINE.value);

        bind(km, KeyModifier.CTRL, KeyCode.D, EditorBuiltinCommand.COPY_LINE_DOWN.value);
        bind(km, KeyModifier.META, KeyCode.D, EditorBuiltinCommand.COPY_LINE_DOWN.value);

        bind(km, KeyModifier.SHIFT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_BELOW.value);
        bind(km, KeyModifier.CTRL | KeyModifier.ALT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_ABOVE.value);
        bind(km, KeyModifier.META | KeyModifier.ALT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_ABOVE.value);

        bind(km, KeyModifier.ALT | KeyModifier.SHIFT, KeyCode.UP,   EditorBuiltinCommand.MOVE_LINE_UP.value);
        bind(km, KeyModifier.ALT | KeyModifier.SHIFT, KeyCode.DOWN, EditorBuiltinCommand.MOVE_LINE_DOWN.value);

        return km;
    }

    /**
     * Sublime Text key bindings.
     */
    public static EditorKeyMap sublime() {
        EditorKeyMap km = new EditorKeyMap();
        addCommonBindings(km);

        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.Z, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.Z, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.CTRL, KeyCode.Y, EditorBuiltinCommand.REDO.value);
        bind(km, KeyModifier.META, KeyCode.Y, EditorBuiltinCommand.REDO.value);

        bind(km, KeyModifier.CTRL, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_BELOW.value);
        bind(km, KeyModifier.META, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_BELOW.value);
        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_ABOVE.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.ENTER, EditorBuiltinCommand.INSERT_LINE_ABOVE.value);

        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.UP,   EditorBuiltinCommand.MOVE_LINE_UP.value);
        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.DOWN, EditorBuiltinCommand.MOVE_LINE_DOWN.value);

        bind(km, KeyModifier.CTRL | KeyModifier.SHIFT, KeyCode.K, EditorBuiltinCommand.DELETE_LINE.value);
        bind(km, KeyModifier.META | KeyModifier.SHIFT, KeyCode.K, EditorBuiltinCommand.DELETE_LINE.value);

        return km;
    }
}
